File: CSharp/Shared/ClientFactory.cs
using System;
using System.IO;

namespace Shared
{
    public static class ClientFactory
    {
        private const string FtpRoot = "sftp";

        public static IClientFactory Create(int transferDelay = 0)
        {
            var rootDir = Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, FtpRoot));
            if (!Directory.Exists(rootDir)) Directory.CreateDirectory(rootDir);
            return new LocalFileClientFactory(rootDir, "", transferDelay);
        }
    }
}

File: FSharp/Shared/ClientFactory.fs
module ClientFactory

    open System.IO
    open SftpClient
    open SshNetClient
    open LocalFileClient

    let private ftpRoot = "sftp"

    let createClientFactory () =
        if not (Directory.Exists(ftpRoot)) then Directory.CreateDirectory(ftpRoot) |> ignore
        LocalFileClientFactory(ftpRoot, "", 0<s/MB>)

    let createClientFactoryWithTransferDelay (transferDelay) =
        LocalFileClientFactory(ftpRoot, "", transferDelay)
